src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests2.cs

```diff
@@ -2596,5 +2596,139 @@ public void M(string s)
+            _ => throw null,
+        };
+        D x5 = i switch // 5
+        {
+            0 => new A(),
+            1 => new B(),
+            2 => new C(),
+            _ => throw null,
+        };
+        D x6 = i switch // 6
+        {
+            0 => 1,
+            1 => 2,
+            _ => throw null,
+        };
+    }
+}
+
+class A : C { }
+class B : C { }
+class D
+{
+    public static implicit operator D(C c) => throw null;
+    public static implicit operator D(short s) => throw null;
+}
+class E
+{
+    public static implicit operator C(E c) => throw null;
+}
+class F
+{
+    public static implicit operator C(F c) => throw null;
+}
+";
+            var comp = CreateCompilation(source, parseOptions: TestOptions.RegularWithPatternCombinators).VerifyDiagnostics(
+                // (11,15): error CS8506: No best type was found for the switch expression.
+                //         _ = i switch // 1
+                Diagnostic(ErrorCode.ERR_SwitchExpressionNoBestType, "switch").WithLocation(11, 15),
+                // (25,18): error CS0029: Cannot implicitly convert type 'E' to 'D'
+                //             0 => new E(), // 3.1
+                Diagnostic(ErrorCode.ERR_NoImplicitConv, "new E()").WithArguments("E", "D").WithLocation(25, 18),
+                // (26,18): error CS0029: Cannot implicitly convert type 'F' to 'D'
+                //             1 => new F(), // 3.2
+                Diagnostic(ErrorCode.ERR_NoImplicitConv, "new F()").WithArguments("F", "D").WithLocation(26, 18)
+                );
+            var tree = comp.SyntaxTrees[0];
+            var model = comp.GetSemanticModel(tree);
+            var switches = tree.GetRoot().DescendantNodes().OfType<SwitchExpressionSyntax>().ToArray();
+            for (int i = 0; i < switches.Length; i++)
+            {
+                var @switch = switches[i];
+                var typeInfo = model.GetTypeInfo(@switch);
+                var conversion = model.GetConversion(@switch);
+                switch (i)
+                {
+                    case 0:
+                        Assert.Null(typeInfo.Type);
+                        Assert.Equal("C", typeInfo.ConvertedType.ToTestDisplayString());
+                        Assert.Equal(ConversionKind.SwitchExpression, conversion.Kind);
+                        break;
+                    case 1:
+                        Assert.Null(typeInfo.Type);
+                        Assert.True(typeInfo.ConvertedType.IsErrorType());
+                        Assert.Equal(ConversionKind.Identity, conversion.Kind);
+                        break;
+                    case 2:
+                        Assert.Null(typeInfo.Type);
+                        Assert.Equal("D", typeInfo.ConvertedType.ToTestDisplayString());
+                        Assert.Equal(ConversionKind.SwitchExpression, conversion.Kind);
+                        break;
+                    case 3:
+                        Assert.True(typeInfo.Type.IsErrorType());
```

Comment:
>Assert.True(typeInfo.Type.IsErrorType()); [](start = 24, length = 41)

Why is this an error type rather than `null`?

Reply:
A switch expression that has no natural type and cannot convert to its target type has an error type as its converted type. In this case there is an (error) conversion from the erroneous switch expression to the type `D`. This is the information about that conversion: it converts from the converted type of the switch expression (which is an error type) to the type which is the target of the assignment (`D`).

---
In reply to: [438548774](https://github.com/dotnet/roslyn/pull/45069#discussion_r438548774) [](ancestors = 438548774)